crates/services/importer/src/importer.rs

```diff
@@ -0,0 +1,241 @@
+                let found = !result.is_not_found();
+                // Because the genesis block is not committed, it should return non found error.
+                // If we find the latest height, something is wrong with the state of the database.
+                if found {
+                    return Err(anyhow!("The wrong state of database during insertion of the genesis block"))
+                }
+                actual_next_height
+            }
+            Consensus::PoA(_) => {
+                let last_db_height = self.database.latest_block_height()?;
+                last_db_height
+                    .checked_add(1u32)
+                    .ok_or_else(|| {
+                        anyhow!("An overflow during the calculation of the next height")
+                    })?
+                    .into()
+            }
+        };
+
+        if expected_next_height != actual_next_height {
+            return Err(anyhow!(
+                "The actual height is {}, when the next expected height is {}",
+                actual_next_height.as_usize(),
+                expected_next_height.as_usize(),
+            ))
+        }
+
+        let db_after_execution = db_tx.as_mut();
+
+        // Importer expects that `UncommittedResult` contains the result of block
+        // execution(It includes the block itself).
+        if db_after_execution.latest_block_height()? != actual_next_height {
+            return Err(anyhow!(
+                "`UncommittedResult` doesn't have the block at height {}",
+                actual_next_height
+            ))
+        }
+
+        db_after_execution
+            .seal_block(&block_id, &result.sealed_block.consensus)?
+            .should_be_unique(&actual_next_height)?;
+
+        // TODO: This should use a proper BMT MMR. Based on peaks stored in the database,
+        //  we need to calculate a new root. The data type that will do that should live
+        //  in the `fuel-core-storage` or `fuel-merkle` crate.
+        let root = ephemeral_merkle_root(
+            vec![*block.header().prev_root(), block_id.into()].iter(),
+        );
+        db_after_execution
+            .insert_block_header_merkle_root(&actual_next_height, &root)?
+            .should_be_unique(&actual_next_height)?;
+
+        db_tx.commit()?;
+
+        let _ = self.broadcast.send(Arc::new(result));
+        Ok(())
+    }
+}
+
+impl<IDatabase, E, V> Importer<IDatabase, E, V>
+where
+    IDatabase: ImporterDatabase,
+    E: Executor,
+    V: BlockVerifier,
+{
+    /// The method validates the `Block` fields, executes the `Block`, and commits it with data from
+    /// `SealedBlock`. More about the commit process in [`Importer::commit_result`].
+    ///
+    /// It validates only the `Block` execution rules and the `Block` fields' validity.
+    /// The validity of the `SealedBlock` and seal information is not the concern of this function.
+    pub async fn execute_and_commit(
```

Comment:
This function should not be async as all it needs to do is execute (Which can only happen on the single execution thread), validate (which is just cpu work) and database writing which is sync.

Reply:
I planned to have `Mutex`/`Semaphore` from `tokio`(async lock) and [spawn_blocking](https://docs.rs/tokio/0.2.25/tokio/task/fn.spawn_blocking.html) for executing/committing(to not handle that on the caller's side + in the future executor will be async for parallel execution). But yeah, let's minimize async for as possible and add it later.

Because right now, either `PoA` or `fuel-core-sync` will import blocks, we shouldn't have cases where we really need to await lock. So, for now, we will have this lock=) 

<img width="796" alt="image" src="https://user-images.githubusercontent.com/18346821/213168971-7ecac90a-98f0-4c58-bbaf-e32917a31761.png">